lightning/src/chain/channelmonitor.rs

```diff
@@ -2054,29 +2237,58 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 			}
 			claimable_outpoints.append(&mut new_outpoints);
 		}
-		if let Some(events) = self.onchain_events_waiting_threshold_conf.remove(&height) {
-			for ev in events {
-				match ev {
-					OnchainEvent::HTLCUpdate { htlc_update } => {
-						log_trace!(logger, "HTLC {} failure update has got enough confirmations to be passed upstream", log_bytes!((htlc_update.1).0));
-						self.pending_monitor_events.push(MonitorEvent::HTLCEvent(HTLCUpdate {
-							payment_hash: htlc_update.1,
-							payment_preimage: None,
-							source: htlc_update.0,
-						}));
-					},
-					OnchainEvent::MaturingOutput { descriptor } => {
-						log_trace!(logger, "Descriptor {} has got enough confirmations to be passed upstream", log_spendable!(descriptor));
-						self.pending_events.push(Event::SpendableOutputs {
-							outputs: vec![descriptor]
-						});
-					}
+
+		// Find which on-chain events have reached their confirmation threshold.
+		let onchain_events_awaiting_threshold_conf =
+			self.onchain_events_awaiting_threshold_conf.drain(..).collect::<Vec<_>>();
+		let mut onchain_events_reaching_threshold_conf = Vec::new();
+		for entry in onchain_events_awaiting_threshold_conf {
+			if entry.has_reached_confirmation_threshold(height) {
+				onchain_events_reaching_threshold_conf.push(entry);
+			} else {
+				self.onchain_events_awaiting_threshold_conf.push(entry);
+			}
+		}
+
+		// Used to check for duplicate HTLC resolutions.
+		#[cfg(debug_assertions)]
+		let unconfirmed_htlcs: Vec<_> = self.onchain_events_awaiting_threshold_conf
+			.iter()
+			.filter_map(|entry| match &entry.event {
+				OnchainEvent::HTLCUpdate { htlc_update } => Some(htlc_update.0.clone()),
+				OnchainEvent::MaturingOutput { .. } => None,
+			})
+			.collect();
+		#[cfg(debug_assertions)]
+		let mut confirmed_htlcs = Vec::new();
+
+		// Produce actionable events from on-chain events having reached their threshold.
+		for entry in onchain_events_reaching_threshold_conf.drain(..) {
+			match entry.event {
+				OnchainEvent::HTLCUpdate { htlc_update } => {
+					// Check for duplicate HTLC resolutions.
+					debug_assert!(unconfirmed_htlcs.iter().find(|&htlc| htlc == &htlc_update.0).is_none());
```

Comment:
Since these are "user feedback assertions", they should have clear error messages - something like "We reached several confirmations on a transaction, while still believing that a conflicting transaction is confirmed. This implies you've failed to call transaction_unconfirmed on some transaction which is no longer in the best chain in a timely manner or otherwise missed relevant transactions or block disconnections.".